src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/OAuth2/OAuthClient.cs

```diff
@@ -0,0 +1,255 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved. 
+// Licensed under the MIT License.
+
+using Microsoft.Identity.Core.Http;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Internal;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Helpers;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Http;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.OAuth2;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Platform;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Core.OAuth2
+{
+    internal class OAuthClient
+    {
+        private const string DeviceAuthHeaderName = "x-ms-PKeyAuth";
+        private const string DeviceAuthHeaderValue = "1.0";
+        private const string WwwAuthenticateHeader = "WWW-Authenticate";
+        private const string PKeyAuthName = "PKeyAuth";
+        internal /* internal for test only */ const string ExtraQueryParamEnvVariable = "ExtraQueryParameter";
+        internal bool Resiliency = false;
+        private readonly RequestContext _requestContext;
+        private readonly IHttpManager _httpManager;
+        private readonly Dictionary<string, string> _headers = new Dictionary<string, string>();
+
+        internal OAuthClient() { }
+
+        public OAuthClient(IHttpManager httpManager, string uri, RequestContext requestContext)
+        {
+            _httpManager = httpManager ?? throw new ArgumentNullException(nameof(httpManager));
+            RequestUri = CheckForExtraQueryParameter(uri);
+            _requestContext = requestContext;
+        }
+
+        public int TimeoutInMilliSeconds { set; get; } = 30000;
+        public IRequestParameters BodyParameters { get; set; }
+        internal string RequestUri { get; set; }
```

Comment:
Please review all the setters - if you set it from the constructor, you don't need to add "set". If you set it only from this class, do interal string Prop { get; private set; }. This greatly helps understanding who is allowed to set this property. Ideally all properties and members are readonly, i.e. only the ctor is allowed to set them.